### PR TITLE
Scale dry air mass on initalization for non-analytic, non-restart simulations 

### PR DESCRIPTION
**Pull request by MiCurry**
This merge adds code to the MPAS dycore which allows the ability to scale the dry air mass for non-analytic and non-restart simulations. To enable dry air mass scaling, the namelist option, mpas_scale_dry_air_mass, which this merge also introduces, must be set to .true. in user_nl_cam.

This merge also adds a routine that can compute the global sum of a one dimensional real array across all tasks. For now, this single real routine is needed, but a future commit can add an additional routine to compute the global sum of a integer array and could add an abstract interface for both functions.

### CODE REVIEW

**Comment by MiCurry** [comment]
This PR is the same PR as #18, execpt it is now up-to-date with `cam_mpas_dev` and points to my new CAM fork, which was deleted so the upstream could be changed to ESCOMP.

File: src/dynamics/mpas/dyn_comp.F90 [hunk 1]
```diff
@@ -1028,6 +1045,69 @@ subroutine read_inidat(dyn_in)
 
    theta_m(:,1:nCellsSolve) = theta(:,1:nCellsSolve) * (1.0_r8 + Rv_over_Rd * tracers(ixqv,:,1:nCellsSolve))
 
+   ! if (.not. analytic_ic_active()) then  ! scale dry-air mass
+   !   surface_integral = cam_mpas_global_sum_real(areaCell(1:nCellsSolve))
+   !   write(iulog,*) subname//': Cell area test value = ', surface_integral
+   !   test_value = sqrt(surface_integral/(4.0_r8*pi))
+   !   write(iulog,*) subname//': earth radius from area = ', test_value
+   ! end if
+
```
Comment (jtruesdal) [inline src/dynamics/mpas/dyn_comp.F90]: Remove dead code

File: src/dynamics/mpas/dyn_comp.F90 [hunk 2]
```diff
@@ -1028,6 +1045,69 @@ subroutine read_inidat(dyn_in)
...
+   call mpas_pool_get_config(domain_ptr % configs, 'config_scale_dry_air_mass', mpas_scale_dry_air_mass)
+
+   if (mpas_scale_dry_air_mass) then
+
+      allocate( p_top(nCellsSolve), preliminary_dry_surface_pressure(nCellsSolve), pm(plev) )
+      ! (1) calculate pressure at the lid
```
Comment (jtruesdal) [inline src/dynamics/mpas/dyn_comp.F90]: I would take this whole section and place it in a new routine.  SE and FV3 named similar routines set_dry_mass.  In addition, it needs to have log output similar to what SE and FV3 provide (ie. initial dry mass, scaled dry mass )

File: src/dynamics/mpas/dyn_comp.F90 [hunk 3]
```diff
@@ -1028,6 +1045,69 @@ subroutine read_inidat(dyn_in)
...
+
+      ! (4) scale dry air density                                                                                                                 
+      ! scaling_ratio = target_avg_dry_surface_pressure/preliminary_avg_dry_surface_pressure
+      scaling_ratio = 1.0
+      rho(:,:) = rho(:,:)*scaling_ratio
+
+      ! (5) reset qv to conserve mass                                                                                                             
+      tracers(ixqv,:,1:nCellsSolve) = tracers(ixqv,:,1:nCellsSolve)/scaling_ratio
+
+      ! (6) integrate down the column to compute full pressure given the density and qv
```
Comment (jtruesdal) [inline src/dynamics/mpas/dyn_comp.F90]: @PeterHjortLauritzen Do we have to worry about dry/wet tracers in this dycore.  ie conserve mass of wet tracers vs conserve mixing ration of dry?

Comment (PeterHjortLauritzen) [reply src/dynamics/mpas/dyn_comp.F90]: All mixing ratios in MPAS (incl. those on the initial condition file) are dry so no.

File: src/dynamics/mpas/dyn_comp.F90 [hunk 4]
```diff
@@ -1028,6 +1045,69 @@ subroutine read_inidat(dyn_in)
...
+      sphere_surface_area = cam_mpas_global_sum_real(areaCell(1:nCellsSolve))
+      preliminary_dry_surface_pressure = cam_mpas_global_sum_real(preliminary_dry_surface_pressure(1:nCellsSolve))
+      preliminary_avg_dry_surface_pressure = preliminary_avg_dry_surface_pressure/sphere_surface_area
+      write(iulog,*) subname//': initial dry globally avg surface pressure (hPa) = ', preliminary_avg_dry_surface_pressure/100.
+
+      ! (4) scale dry air density                                                                                                                 
+      ! scaling_ratio = target_avg_dry_surface_pressure/preliminary_avg_dry_surface_pressure
+      scaling_ratio = 1.0
+      rho(:,:) = rho(:,:)*scaling_ratio
+
```
Comment (jtruesdal) [inline src/dynamics/mpas/dyn_comp.F90]: @PeterHjortLauritzen how about adding a dry mass reference pressure (98288.0) to physconst.F90 that all the dycores can use for scaling.

Comment (PeterHjortLauritzen) [reply src/dynamics/mpas/dyn_comp.F90]: Yes and higher value without Topo:

      initial_global_ave_dry_ps = 98288.0_r8
      if (.not. associated(fh_topo)) then
         initial_global_ave_dry_ps = 101325._r8 - 245._r8

File: src/dynamics/mpas/dyn_comp.F90 (hunk 2 above)
Comment (MiCurry) [reply src/dynamics/mpas/dyn_comp.F90]: Sounds good. I'll just create a new function in dyn_comp.

Comment (jtruesdal) [reply src/dynamics/mpas/dyn_comp.F90]: Great. Feel free to use the FV3 dyn_comp.F90 as a template for types of logging messages.  They will help us validate the additional functionality.  Sounds like Peter is OK with adding these pressures in physconst.F90, maybe ps_dry_topo, ps_dry_notopo ! CESM reference dry mass pressure with/without topography.  I can change them during the PR to cam_development if needed.

File: src/dynamics/mpas/dyn_comp.F90 [hunk 5]
```diff
@@ -1028,6 +1045,69 @@ subroutine read_inidat(dyn_in)
...
+      ! (6) integrate down the column to compute full pressure given the density and qv                                                           
+
+      do i=1,nCellsSolve
+         pm(plev) = p_top(i) + 0.5_r8*(zint(plev+1,i)-zint(plev,i))*gravity*rho(plev,i)*(1.0_r8+tracers(ixqv,plev,i))
+         do k=plev-1,1,-1
+            pm(k) = pm(k+1) + 0.5_r8*(zint(k+2,i)-zint(k+1,i))*gravity*rho(k+1,i)*(1.0_r8+tracers(ixqv,k+1,i)) &
+                            + 0.5_r8*(zint(k+1,i)-zint(k  ,i))*gravity*rho(k  ,i)*(1.0_r8+tracers(ixqv,k  ,i))
+         end do
+
+      ! (7) theta_m from the state equation, compute rho_zz and theta while we are here
```
Comment (MiCurry) [inline src/dynamics/mpas/dyn_comp.F90]: I believe this comment should be: "Compute theta_m from the state equation", correct @skamaroc?

**Comment by MiCurry** [comment]
> Great. Feel free to use the FV3 dyn_comp.F90 as a template for types of logging messages. They will help us validate the additional functionality. Sounds like Peter is OK with adding these pressures in physconst.F90, maybe ps_dry_topo, ps_dry_notopo ! CESM reference dry mass pressure with/without topography. I can change them during the PR to cam_development if needed.

Sounds good. I moved this code to a new subroutine today, removed the dead code and added the log messages in your branch, I'll double check and see if the log messages are the same with the other dycores and then I'll add an additional commit which adds to `physconst.F90`.

**Comment by MiCurry** [comment]
@jtruesdal @PeterHjortLauritzen I have just pushed four additional commits that: 1.) Moves the dry mass scaling to a new function, set_dry_mass, 2.) Add the physical constants mentioned above to physconst.F90 and uses them in the MPAS dycore and 3.) adds in additional log messages, which includes calculating the scaled dry air mass (taken from John's Fork).

Please feel free to give it another review. Are there any other log messages you think should be added?

File: src/dynamics/mpas/dyn_comp.F90 (hunk 1 above)
Comment (MiCurry) [reply src/dynamics/mpas/dyn_comp.F90]: removed

File: src/dynamics/mpas/dyn_comp.F90 (hunk 2 above)
Comment (MiCurry) [reply src/dynamics/mpas/dyn_comp.F90]: Moved to a new function. And added ps_dry_topo/notopo to physconst.F90.

**Comment by MiCurry** [comment]
@jtruesdal Is there a namelist group you recommend adding `scale_dry_air_mass` so that it can be easily accessible by all dycores? Perhaps, `ref_pres_nl`?

**Comment by jtruesdal** [comment]
We have two that I think would work 1) phys_ctl_nl   2) air_composition_nl which specifies wet/dry air compositions. 
 @PeterHjortLauritzen what would you like?

**Comment by jtruesdal** [comment]
Actually ref_pres_nl sounds like the natural place for this.  I'm behind the curve, I didn't even know that existed.

**Comment by PeterHjortLauritzen** [comment]
I am not sure ref_pres_nl is the right place when looking at the variables in that namelist. It almost seems like we need an initialization namelist?

**Comment by MiCurry** [comment]
@PeterHjortLauritzen @jtruesdal There is the `cam_initfiles_nl`, which has `pertlim`, which perturbs the temperature up by a given amount. Its also defaults to 0.0.

**Comment by PeterHjortLauritzen** [comment]
Great suggestion. What do you think @jtruesdal?

**Comment by jtruesdal** [comment]
Thats it!

**Comment by MiCurry** [comment]
~Okay, @jtruesdal and @PeterHjortLauritzen, I just pushed an additional commit which adds the nl option, scale_dry_air_mass, and changes the logic of the calling code for the `set_dry_mass`.~

~@PeterHjortLauritzen, I updated the code from your pseudocode so that there is only one call to`set_dry_mass` (rather than two calls in separate locations).~

~Does the namelist description for `scale_dry_air_mass` look appropriate to you?~

Actually looks like this doesn't run. I'll update y'all with changes with it tomorrow. @jtruesdal is it failing to run because I'm reading the namelist wrong?

**Comment by jtruesdal** [comment]
@MiCurry Looks like you forgot to add scale_dry_air_mass to namelist /cam_initfiles_nl/ in cam_initfiles.F90.  It will need to be defined as well as mpi_bcast like the rest.

jt

**Comment by MiCurry** [comment]
Okay, sorry for the delay @jtruesdal and @PeterHjortLauritzen, I just pushed an update to this branch which, instead of reading the cam_initfiles_nl in the MPAS dycore, adds `scale_dry_air_mass` namelist variable as a public variable in `cam_initfiles.F90`, which each dycore can access on its on with `use`.

I also added units to `ps_dry_topo/notopo` in `physconts.F90`, I did so with a `git amend`, so if you have this branch checked out, you'll need to reset your branch to match my remote, which you can do by:

```
git fetch curry #Fetch my remote
git checkout cam-mpas/mass_scaling # If its not already checked out
git reset --hard curry/cam-mpas/mass_scaling # Update your branch to match my remote
```

I tested this branch successfully with different values of scale_dry_air_mass (i.e. < 0, 0.0, ~98000.0)

Let me know if you have any questions or problems.

**Comment by jtruesdal** [comment]
Thanks Miles I'm testing it now.

jt

**Pull request merged**